File: aat-android/src/main/java/ch/bailu/aat/util/GpxPointPoi.java
package ch.bailu.aat.util;


import org.mapsforge.core.model.Tag;
import org.mapsforge.poi.storage.PointOfInterest;

import java.util.Set;

import ch.bailu.aat_lib.gpx.attributes.GpxAttributes;
import ch.bailu.aat_lib.gpx.attributes.GpxAttributesStatic;
import ch.bailu.aat_lib.gpx.attributes.Keys;
import ch.bailu.aat_lib.gpx.interfaces.GpxPointInterface;

public class GpxPointPoi implements GpxPointInterface {

    private final static int KEY_ELE = Keys.toIndex("ele");
    private final static int KEY_ALTITUDE = Keys.toIndex("altitude");

    final PointOfInterest poi;
    final GpxAttributes attr;

    public GpxPointPoi(PointOfInterest p) {
        poi = p;
        attr = toAttributes(p.getTags());
    }

    @Override
    public double getAltitude() {
        if (attr.hasKey(KEY_ELE))
            return Double.parseDouble(attr.get(KEY_ELE));

        if (attr.hasKey(KEY_ALTITUDE))
            return Double.parseDouble(attr.get(KEY_ALTITUDE));

        return 0d;
    }

    private GpxAttributes toAttributes(Set<Tag> tags) {
        final GpxAttributesStatic.Tag[] newTags =
                new GpxAttributesStatic.Tag[tags.size()];

        int i=0;
        for(Tag tag : tags) {
            newTags[i++] =
                    new GpxAttributesStatic.Tag(Keys.toIndex(tag.key), tag.value);
        }
        return new GpxAttributesStatic(newTags);
    }

    @Override
    public double getLongitude() {
        return poi.getLongitude();
    }

    @Override
    public double getLatitude() {
        return poi.getLatitude();
    }

    @Override
    public long getTimeStamp() {
        return 0;
    }

    @Override
    public GpxAttributes getAttributes() {
        return attr;
    }

    @Override
    public int getLatitudeE6() {
        return (int) (getLatitude() * 1e6);
    }

    @Override
    public int getLongitudeE6() {
        return (int) (getLongitude() * 1e6);
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/api/poi/GpxPointPoi.kt
package ch.bailu.aat_lib.api.poi

import ch.bailu.aat_lib.gpx.attributes.GpxAttributes
import ch.bailu.aat_lib.gpx.attributes.GpxAttributesStatic
import ch.bailu.aat_lib.gpx.attributes.Keys.Companion.toIndex
import ch.bailu.aat_lib.gpx.interfaces.GpxPointInterface
import ch.bailu.aat_lib.service.elevation.ElevationProvider
import org.mapsforge.core.model.Tag
import org.mapsforge.poi.storage.PointOfInterest


class GpxPointPoi(val poi: PointOfInterest) : GpxPointInterface {
    private val attributes: GpxAttributes = toAttributes(poi.tags)

    override fun getAltitude(): Float {
        if (attributes.hasKey(KEY_ELE)) return attributes[KEY_ELE].toFloat()

        if (attributes.hasKey(KEY_ALTITUDE)) return attributes[KEY_ALTITUDE].toFloat()

        return ElevationProvider.NULL_ALTITUDE
    }

    override fun getLongitude(): Double {
        return poi.longitude
    }

    override fun getLatitude(): Double {
        return poi.latitude
    }

    override fun getTimeStamp(): Long {
        return 0
    }

    override fun getAttributes(): GpxAttributes {
        return attributes
    }

    override fun getLatitudeE6(): Int {
        return (getLatitude() * 1e6).toInt()
    }

    override fun getLongitudeE6(): Int {
        return (getLongitude() * 1e6).toInt()
    }

    companion object {
        private val KEY_ELE = toIndex("ele")
        private val KEY_ALTITUDE = toIndex("altitude")

        private fun toAttributes(tags: Set<Tag>): GpxAttributes {
            val attributes = ArrayList<GpxAttributesStatic.Tag>()
            tags.forEach {
                attributes.add(GpxAttributesStatic.Tag(toIndex(it.key), it.value))
            }
            return GpxAttributesStatic(attributes.toTypedArray())
        }
    }
}
